nixos/modules/misc/documentation.nix

```diff
@@ -48,45 +47,46 @@ let
         };
         scrubDerivations = namePrefix: pkgSet: mapAttrs
           (name: value:
-            let wholeName = "${namePrefix}.${name}"; in
-            if isAttrs value then
+            let
+              wholeName = "${namePrefix}.${name}";
+              guard = lib.warn "Attempt to evaluate package ${wholeName} in option documentation; this is not supported and will eventually be an error. Use `mkPackageOption` or `literalExpression` instead.";
+            in if isAttrs value then
               scrubDerivations wholeName value
-              // (optionalAttrs (isDerivation value) { outPath = "\${${wholeName}}"; })
+              // optionalAttrs (isDerivation value) {
+                outPath = guard "\${${wholeName}}";
+                drvPath = guard drvPath;
+              }
             else value
           )
           pkgSet;
       in scrubbedEval.options;
 
     baseOptionsJSON =
-      let
-        filter =
-          builtins.filterSource
-            (n: t:
-              cleanSourceFilter n t
-              && (t == "directory" -> baseNameOf n != "tests")
-              && (t == "file" -> hasSuffix ".nix" n)
-            );
-      in
         pkgs.runCommand "lazy-options.json" {
-          libPath = filter (pkgs.path + "/lib");
-          pkgsLibPath = filter (pkgs.path + "/pkgs/pkgs-lib");
-          nixosPath = filter (pkgs.path + "/nixos");
+          filteredNixpkgs = builtins.path {
+            name = "filtered-nixpkgs";
+            path = pkgs.path;
+            filter = let dirs = [ "lib/" "pkgs/pkgs-lib/" "nixos/" ]; in
+              path: type: let p = lib.removePrefix (toString pkgs.path + "/") (toString path); in
+              cleanSourceFilter path type && (
+                if type == "directory" then
+                  baseNameOf path != "tests" && lib.any (dir: lib.hasPrefix dir (p + "/") || lib.hasPrefix (p + "/") dir) dirs
+                else
+                  p == ".version" || lib.any (dir: lib.hasPrefix dir p) dirs);
+          };
```

Comment:
The `.nix` filtering never worked, because the check was `t == "file"` instead of `t == "regular"`, and in fact excluding non-nix files isn't possible currently because of e.g. https://github.com/NixOS/nixpkgs/blob/cecc553095ea244e516003f7e0714669afb4af70/nixos/modules/services/system/nscd.nix#L57.

We could still make it work by adding a `defaultText` there, or marking that module non-cacheable, or we could just filter out directories named `doc`

Reply:
I think we should avoid bothering module authors with this implementation detail as much as possible, so I went with excluding `doc`. The resulting store path weighs 13.5 MiB, of which 12.6 MiB are .nix files.